### PR TITLE
Fix bug in negative string section

### LINKED ISSUE #4741: Bug in negative string section
```fortran
$ cat examples/expr2.f90 
program expr2
implicit none

character(len=5) :: x

x = "hello"

print *, x(1:-1)

end program
```

```console
(lf) ubaid@DESKTOP-TEN39M0:~/projects/lfortran$ gfortran examples/expr2.f90 
(lf) ubaid@DESKTOP-TEN39M0:~/projects/lfortran$ ./a.out 
 
(lf) ubaid@DESKTOP-TEN39M0:~/projects/lfortran$ lfortran_in_main examples/expr2.f90 
hell
```

### PR DESCRIPTION
**Pull request by Shaikh-Ubaid**
fixes #4741

### CODE REVIEW

File: src/libasr/runtime/lfortran_intrinsics.c [hunk 1]
```diff
@@ -2179,8 +2179,15 @@ LFORTRAN_API char* _lfortran_str_slice(char* s, int32_t idx1, int32_t idx2, int3
...
         exit(1);
     }
+#ifdef COMPILER_LFORTRAN
+    if (idx1 < 0) {
+        printf("Substring start index is negative\n");
+        exit(1);
+    }
+#else
     idx1 = idx1 < 0 ? idx1 + s_len : idx1;
     idx2 = idx2 < 0 ? idx2 + s_len : idx2;
```
Comment (Shaikh-Ubaid) [inline src/libasr/runtime/lfortran_intrinsics.c:2189]: Python supports negative indexing in string slicing, but fortran does not. So we need some special handling based on the compiler using this runtime library. This is achieved by enabling the macro `COMPILER_LFORTRAN` when we are compiling the `lfortran` project. We can later similarly have `COMPILER_LPYTHON` macro later as needed.

**Comment by certik** [comment]
This one is tricky. It seems a cleaner solution is to store this in ASR with a flag if negative indexing is allowed, and then in the backend/runtime we act accordingly.

**Comment by Shaikh-Ubaid** [comment]
Thanks @gxyd and @certik. I will update to support negative values at runtime.

>Is it intentional to create this PR against "simplifer_pass" branch?

Yes, it is intentional. Although the bug is also present in `main` (in a slightly varied form), fixing it in the simplifier pass by default fixes it in the main. Plus we get to uncomment a simplifier failing test.

>As I think the bug in "simplifier pass" branch is a little different, and it gets fixed when changes in https://github.com/lfortran/lfortran/pull/4760 are made.

Thanks, I shared my ideas at https://github.com/lfortran/lfortran/pull/4760. We should see what the ideal aim of simplifier pass is and proceed accordingly.

**Comment by Shaikh-Ubaid** [comment]
>It seems a cleaner solution is to store this in ASR with a flag if negative indexing is allowed, and then in the backend/runtime we act accordingly.

I was making this change, but then had the following second thought. It seems string section `s(a:b)` is a very common operation supported by multiple languages (fortran, python, c). I think if we do the check for `is negative indexing supported` at runtime, then we might possibly slow down the string section operation for all the different lcompilers. @certik can you confirm if checking for negative indexing at runtime is a good approach or if the one currently used in this PR is good?

**Comment by certik** [comment]
Yes, we don't want to slow down runtime, and especially not for common operations. The current implementation at runtime is not fast anyway, it's already doing all kinds of runtime checks, etc. To make it really fast, we would need to implement using the same approach as intrinsics: expanding in ASR itself. But that we can do later.

For now it seems adding an argument "allow_negative_indexing" into the runtime function will not significantly slow it down, since right now the implementation is mainly focused on doing the correct operation, but not necessarily on performance.

In ASR I think it won't slow things down by adding this flag, nor the LLVM backend will get significantly slower by passing through this flag.

And then later, if we want to, we can provide a faster implementation.

So it seems overall adding this flag into ASR is the right approach. Let me know what you think.

**Comment by Shaikh-Ubaid** [comment]
>So it seems overall adding this flag into ASR is the right approach. Let me know what you think.

Sounds good. Created a new PR https://github.com/lfortran/lfortran/pull/4869 towards `main` branch. Closing this PR in favor of it.

**Pull request closed without merging**